### DISCUSSION TITLE
Log of played tracks

### DISCUSSION THREAD

**Discussion by jej** [quodlibet/quodlibet#3863] [Q&A] [answered]
Great player, thanks!
I was looking for a list of played songs (something where I can see all the past played songs with played date). Something like "Logs Radio Internet".
Does it exists in quodlibet or as a plugin? I didn't find that.
Thanks,
J.

**Comment by declension** [comment, marked as answer]
No, but lots of people use Scrobblers (e.g. last.fm) for this kind of thing (esp as they work across devices). These are supported by the _AudioScrobbler Submission_ plugin and the _ListenBrainz Submission_ plugin (which I've just noticed doesn't appear - filed #3864)

However yes it would be nicer if there was a QL-based widget for this song history.

Note that certain newer modes (persistent mode) for the queue has _some_ similarities

See also: #894, #1407, #3160 and some others

**Reply by jej** [reply]
Ok thanks. I was thinking about a local solution, just a simple log that could be browsed. Maybe I can try to code a small plug-in, if I have some guidance to do that. Sometime I am a bit bored with django so it could be fun to make some bushwhacking diversions :)

Maybe an event plug-in could feed a playlist, shorted to a maximum elements number. Or it could save this in the database. I see that next/back button can browse the tracks played so I guess the played tracks are already somewhere...

I don't know, I have so few background with quodlibet I cannot decide of the best strategy. I spent most of the time to listen music not code quodlibet! So if you have some advices, just tell me.
J.

**Comment by jej** [comment]
I have something almost working, the problem is the playlist is not refreshed when a song is added. What is the trick?

```
from quodlibet.plugins.events import EventPlugin
from quodlibet import _, app, browsers
from quodlibet.library.playlist import PlaylistLibrary

LOG_PLAYLIST = "Played tracks"

class LogToPlaylist(EventPlugin):
    PLUGIN_ID = "logtoplaylist"
    VERSION = "0.1"
    PLUGIN_NAME = _("Log to playlist")
    PLUGIN_DESC = _("Log played tracks in a playlist")

    def __init__(self):
        super().__init__()

        pl_lib = PlaylistLibrary(app.library)
        if pl_lib.has_key(LOG_PLAYLIST):
            self.log_playlist = pl_lib.get(LOG_PLAYLIST)
        else:
            self.log_playlist = pl_lib.create(LOG_PLAYLIST)
            self.log_playlist.write()

        print(self.log_playlist)

    def plugin_on_song_started(self, song):
        if song is not None:
            self.log_playlist.append(song)
            self.log_playlist.finalize()
            self.log_playlist.write()
            #browsers.PlaylistsBrowser.refresh_all():


```
